staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,633 @@
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.Packets = pkts[removeIdx:]
+		qs.DecrementPackets(removeIdx)
+	}
+}
+
+// DecrementPackets decreases the # of packets for the QueueSetImpl w/ lock
+func (qs *QueueSetImpl) DecrementPackets(i int) {
+	qs.numPacketsEnqueued -= i
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishPacket method invoked after service)
+func (qs *QueueSetImpl) GetRequestsExecuting() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+func shuffleDealAndPick(v, nq uint64,
+	lengthOfQueue func(int) int,
+	mr func(int /*in [0, nq-1]*/) int, /*in [0, numQueues-1] and excluding previously determined members of I*/
+	nRem, minLen, bestIdx int) int {
+	if nRem < 1 {
+		return bestIdx
+	}
+	vNext := v / nq
+	ai := int(v - nq*vNext)
+	ii := mr(ai)
+	mrNext := func(a int /*in [0, nq-2]*/) int /*in [0, numQueues-1] and excluding I[0], I[1], ... ii*/ {
+		if a < ai {
+			return mr(a)
+		}
+		return mr(a + 1)
+	}
+	lenI := lengthOfQueue(ii)
+	if lenI < minLen {
+		minLen = lenI
+		bestIdx = ii
+	}
+	return shuffleDealAndPick(vNext, nq-1, lengthOfQueue, mrNext, nRem-1, minLen, bestIdx)
+}
+
+// ChooseQueueIdx uses shuffle sharding to select an queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (qs *QueueSetImpl) ChooseQueueIdx(hashValue uint64, handSize int) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
+	// verify that makes sense...
+
+	// desiredNumQueues is used here instead of numQueues to omit quiesce queues
+	return shuffleDealAndPick(hashValue, uint64(qs.desiredNumQueues),
+		func(idx int) int { return len(qs.queues[idx].Packets) },
+		func(i int) int { return i }, handSize, math.MaxInt32, -1)
+}
+
+// rejectOrEnqueue rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *QueueSetImpl) rejectOrEnqueue(packet *Packet) bool {
+	queue := packet.Queue
+	curQueueLength := len(queue.Packets)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.GetRequestsExecuting() >= qs.concurrencyLimit &&
+		curQueueLength >= qs.queueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(packet)
+	return true
+}
+
+// enqueues a packet into an QueueSetImpl
+func (qs *QueueSetImpl) enqueue(packet *Packet) {
+
+	queue := packet.Queue
+	queue.Enqueue(packet)
+	qs.updateQueueVirStartTime(packet, queue)
+	qs.numPacketsEnqueued++
+}
+
+// Enqueue enqueues a packet directly into an QueueSetImpl w/ no restriction
+func (qs *QueueSetImpl) Enqueue(packet *Packet) bool {
```

Comment:
This function is here only for testing, right?
I suggest testing through the QueueSet interface, as in https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fq_test.go , as much as possible rather than using special entry methods.  Of course, for testing that is impossible to accomplish that way then another way must be used.